2주차 Spring Web으로 ToDo REST API 만들기

eunhasoo opened the pull request:
안녕하세요, 트레이너님.
일단 동작하도록 중점을 두고 구현해보았고 차근차근 리팩토링 해보겠습니다.

+) API 테스트는 완료했는데 E2E 테스트는 아래와 같은 내용으로 모두 실패하네요. 
m1 맥북이라 그런건지 모르겠는데... chromium도 깔아보고 이것저것 해결해보려다가 실패했습니다 ㅠㅠ
![image](https://user-images.githubusercontent.com/44668991/190983889-e46df077-ae88-43cd-86a8-d0a03501a812.png)

eunhasoo committed:
Feat: Task 도메인 추가

eunhasoo committed:
Feat: TaskRepository 추가

eunhasoo committed:
Feat: TaskController 추가

eunhasoo committed:
Fix: deleteById 메소드 Java doc 수정

johngrib commented:
> 안녕하세요, 트레이너님. 일단 동작하도록 중점을 두고 구현해보았고 차근차근 리팩토링 해보겠습니다.
> 
> +) API 테스트는 완료했는데 E2E 테스트는 아래와 같은 내용으로 모두 실패하네요. m1 맥북이라 그런건지 모르겠는데... chromium도 깔아보고 이것저것 해결해보려다가 실패했습니다 ㅠㅠ ![image](https://user-images.githubusercontent.com/44668991/190983889-e46df077-ae88-43cd-86a8-d0a03501a812.png)

메시지를 읽어보니 Puppeteer 와 관련된 문제 같은데 윤석님께 문의해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,23 @@
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
```
추가 과제입니다. `setId`, `setTitle` 메소드와 같은 setter 메소드를 정의하는 것이 금지되었다고 생각하고 `Task` 클래스를 재설계해보세요.

eunhasoo replied:
빌더 패턴을 적용하고 title을 변경하는 메소드 이름을 updateTitle으로 변경했어요.

johngrib replied:
좋습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,68 @@
+@CrossOrigin
+@RequestMapping("/tasks")
+@RestController
+public class TaskController {
```
오 너무 잘 하셨는데요? 추가 과제를 드리겠습니다. 조회 메소드만 골라서 `TaskQueryController`로, 수정/저장/삭제 메소드만 골라서 `TaskCommandController`로 분리해 보세요. 그리고 이렇게 분리했을 때 어떤 장단점이 있는 것 같은지에 대해서도 이야기해보세요.

eunhasoo replied:
명령과 쿼리가 분리되도록 반영했어요. Command, Query 키워드로 조사하다가 CQRS가 나와서 [여기](https://eunhasoo.notion.site/CQRS-243a14a9154a4b87af3e8bd0ad57875a)에 정리해보았습니다.

johngrib replied:
좋습니다. 굉장히 빠르게 추가 과제를 해내시면서 문서까지 잘 정리하시네요!

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Repository
```
Repository가 어떤 의미를 갖는 단어인지에 대해 설명하는 글을 작성해 봅시다.

eunhasoo replied:
`Repository` 용어에 대해 조사해보라는 의미로 생각하고 정리해봤습니다.
혹시 다른 의미로 말씀하신 거라면 알려주세요!

### 일반적인 용어의 Repository

데이터가 조직적인 방식으로 저장되고 유지되는 중앙 집중화된 장소를 의미한다.

저장소는 다양한 기능을 제공할 수 있다. 네트워크를 통할 필요없이 사용자가 직접적으로 접근할 수 있도록 하거나, 접근이나 배포를 위해 특정한 데이터베이스, 파일, 문서를 저장하는 장소가 될 수도 있다.

### Java에서의 Repository

데이터 소스에 접근하는데 필요한 논리를 캡슐화하는 클래스 또는 구성요소를 의미한다. 공통 데이터 접근 기능을 중앙 집중화하여 유지 보수성을 개선하고, 도메인 모델 계층으로부터 데이터베이스에 접근하는 데 사용하는 인프라 또는 기술을 분리한다.

### Linux에서의 Repository

소프트웨어의 중앙 데이터베이스로 소프트웨어 패키지가 저장되고 관리되며, 공개적으로 접근할 수 있는 장소이다. 운영체제 업데이트나 어플리케이션 설치 등을 수행할 수 있다.

### Github에서의 Repository

프로젝트의 모든 파일과 각 파일의 개정 내역을 갖는다. Repository 내에서 프로젝트 작업물을 관리하고 논의할 수 있다.

johngrib replied:
좋습니다. 좀 더 심화된 내용을 알고 싶으시다면 에릭 에반스의 "도메인 주도 설계"라는 책이 있습니다. 이 책을 구하셔서 "repository"가 나오는 모든 페이지를 읽어보세요. 그리고 내용도 정리해서 글로 작성해 보세요. 좋은 공부가 될 거에요.

eunhasoo replied:
추천 감사합니다! 조금 어려운 책이라고 들었는데 한 번 찾아서 읽어보겠습니다 :)

johngrib reviewed (commented):
와! 월요일인데도 과제를 너무 잘 하셨는데요? 딱히 리뷰를 드릴 것이 없을 정도여서 추가 과제를 드렸습니다.

eunhasoo committed:
Refactor: setter 사용하지 않도록 수정

- Builder 패턴 적용
- set 대신 변경 작업이 나타나는 메소드명 적용

eunhasoo committed:
Refactor: Task 관련 명령과 쿼리 로직을 분리

- TaskController를 TaskQueryController와 TaskCommandController로 분리
- 예외를 처리할 TaskExceptionHandler 생성

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskCommandController.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.util.Optional;
+
+/**
+ * 할 일의 생성/수정/삭제 로직을 처리
+ */
```
이렇게 해보세요.
```suggestion
/**
 * 할 일의 생성/수정/삭제 처리를 담당합니다.
 */
```

eunhasoo replied:
반영했습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskCommandController.java:
```diff
@@ -0,0 +1,51 @@
+@RestController
+@RequestMapping("/tasks")
+@CrossOrigin
+public class TaskCommandController {
```
이 클래스의 모든 public 메소드에 JavaDoc 주석을 작성해 보세요.

eunhasoo replied:
반영했습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskCommandController.java:
```diff
@@ -0,0 +1,51 @@
+    public void deleteTask(@PathVariable Long id) {
+        Optional<Task> deletedTask = taskRepository.deleteById(id);
+        if (deletedTask.isEmpty()) {
+            throw new IllegalArgumentException(id + "에 해당하는 Task가 없어 삭제하지 못했습니다");
```
에러 메시지를 잘 작성하셨습니다. 한편 삭제에 성공했다면 성공했다고 응답하는 방법도 고려해 보세요.

eunhasoo replied:
DELETE API가 성공했을 때 응답 방법을 찾아보니
1. 추가 정보가 필요하지 않은 경우 `204(NO_CONTENT)`와 함께 빈 바디로 응답하기
2. 추가 정보가 필요한 경우 `200(OK)`와 함께 삭제된 데이터 정보를 담아 응답하기

위 두 가지 방법으로 주로 구현하는 것 같습니다.

204번 코드를 사용하면 바디에 데이터를 담아도 무시하고 빈 바디로 응답이 생성되네요.
200번 코드를 사용하면 성공했다는 응답 정보를 추가로 보낼 수 있지만 현재 프로젝트의 API 테스트는 실패하게 되어서 일단 204번으로 응답하도록 그대로 두었습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskQueryController.java:
```diff
@@ -0,0 +1,37 @@
+    @GetMapping("/{id}")
+    public Task getTask(@PathVariable Long id) {
+        return taskRepository.findById(id)
+                .orElseThrow(() -> new IllegalArgumentException(id + "에 해당하는 Task가 존재하지 않습니다."));
```
"~이 존재하지 않습니다."는 상당히 강한 표현입니다. 지금은 로직이 단순해서 생각하기 어렵지만 데이터가 분명히 존재하는데도 버그로 인해 데이터를 찾지 못하거나, 천재지변이나 전력 공급의 문제라거나 하필 우연히 DB에 뭔가 문제가 생겨서 조회에 실패하거나 하는 경우도 발생할 수 있습니다. 그래서 보통은 이렇게 에러 메시지를 작성하곤 하죠.

"~를 찾지 못했습니다."

아마 이런 스타일의 에러 메시지를 그동안 컴퓨터를 사용해오면서 여러 차례 보셨을 거에요.

eunhasoo replied:
오 미처 생각하지 못했던 문제네요! 에러 메시지를 작성할 때는 강한 표현은 피하는 게 좋겠어요. 수정 완료했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/exceptionhandler/TaskExceptionHandler.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+/**
+ * 예외가 발생하면 알맞은 응답을 보내도록 처리
```
모든 예외를 핸들링하지 않고 할 일과 관련된 예외를 처리한다는 것을 JavaDoc에 표현해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,83 @@
+        return createdTask;
+    }
+
+    /**
+     * 기존의 Task를 아이디로 찾아 수정합니다.
+     * @param id 수정할 Task 아이디
+     * @param task 수정할 Task
+     * @return 수정 완료된 Task
+     */
+    public Task update(Long id, Task task) {
```
모든 JavaDoc을 거의 완벽에 가깝게 작성하셨지만 한 가지 생각해볼만한 문제가 있습니다. 바로 `Task` 클래스의 이름이 변경되는 경우죠. 만약 `Task` 클래스의 이름을 `Todo`로 바꾸게 된다면 이 주석과 다른 JavaDoc 주석에 있는 모든 `Task`도 같이 바꿔야 하지 않을까요?

eunhasoo replied:
클래스명이 바뀐다면 기존처럼 작성한 주석이 많을수록 변경하기 힘들어질 것 같아요. 주석도 유지보수성을 생각하며 작성해야겠다는 생각이 드네요. Task 대신 할 일이라는 도메인으로 바꿔봤습니다.
그런데 `@param task 수정할 할 일`과 같이 파라미터에 쓰이는 변수명에 대한 부분은 어쩔 수 없이 변경이 일어날 때 바꿔줘야 하는 부분일까요? task를 data 같은 단어로 바꾸기에는 뭔가 범위가 커지는 느낌이 들어서 애매하네요.

johngrib replied:
`task` 같은 변수명은 IDE의 리팩토링 도구로 이름을 바꾸면 같이 바뀔 거에요. 하지만 `수정할 Task`에서 `Task`의 변경은 지원이 안 될 겁니다. 여기에서 생각해볼 점은 주석이 클래스의 이름에 의존하고 있다는 사실이에요. 의존이 생겨서 변경이 골치아파지는 문제가 발생하는 거죠. 객체지향 원칙이 심지어 주석에도 적용된다고 볼 수 있는 흥미로운 사례입니다. 이건 경험과 감으로 해나가는 수 밖에 없는데요, 저는 보통 이런 경우엔 그냥 `Task` 처럼 클래스 이름을 그냥 쓰지 않고 한국어로 `할 일` 정도로 절충해 주는 편입니다.

eunhasoo replied:
그렇군요! 아무래도 클래스명보다는 도메인명이 변하지 않을 확률이 더 크니까 좋은 습관인 것 같아요. 저도 앞으로 주석 작성할 때 참고하겠습니다 :)

eunhasoo committed:
Docs: Java Doc 추가 및 수정

- TaskCommandController, TaskQueryController 클래스 설명 주석 수정
- TaskCommandController 메소드에 주석 추가

eunhasoo committed:
Refactor: 조회 실패시 에러 메시지 수정

- 없거나 존재하지 않는다는 표현 대신 `찾지 못했다`로 변경

eunhasoo committed:
Docs: 예외 핸들러 주석 수정

- 할 일에 관한 작업중 발생하는 예외를 한정적으로 처리함을 나타내도록 변경

eunhasoo committed:
Docs: TaskRepository 주석 추가 및 수정

- 클래스 상단에 Java doc 추가
- 주석에 불변성을 갖는 도메인이 드러나도록 수정

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,53 @@
+        return title;
+    }
+
+    /**
+     * title을 업데이트 합니다.
+     * @param title 업데이트 할 title
+     */
+    public void updateTitle(String title) {
+        this.title = title;
+    }
```
너무 잘 하고 계셔서 추가 과제를 또 드리고 싶네요. `update` 개념을 `Task` 클래스에서 금지해 보세요. 즉, `Task` 클래스를 불변 클래스로 만들어 보세요.

eunhasoo replied:
불변 클래스로 만들고 이펙티브 자바에 있는 내용을 정리했습니다.

불변 클래스는 인스턴스의 내부 값을 수정할 수 없는 클래스로, 자바에서 제공하는 `Integer`이나 `String` 클래스가 대표적인 예이다.
불변 클래스를 만드려면 아래 다섯가지 규칙을 지키면 된다.
1. 객체의 상태를 변경하는 메서드를 제공하지 않는다
2. 클래스를 확장할 수 없도록 한다
3. 모든 필드를 final으로 선언한다
4. 모든 필드를 private으로 선언한다
5. 자신 외에는 내부의 가변 컴포넌트에 접근할 수 없도록 한다

불변 객체를 사용할 때의 장점은 값을 반환하거나 다른 함수의 인수로 넘겨줄 때 방어적 복사를 할 필요가 없으며 멀티 쓰레드 환경에서 동기화 처리가 필요 없다는 점 등이 있다.
불변 클래스의 단점은 값이 다르면 반드시 독립된 객체로 만들어야 한다는 것이다. 값이 많을수록 비용이 많이 들 수 있다. 100만 비트 `BigInteger` 변수에 하위 비트 하나의 변경이 필요한 경우를 예시로 들 수 있다.

```
BigInteger moby = ...;
moby = moby.flipBit(0);
```

flipBit 메서드는 100만 비트 길이의 새로운 BigInteger 인스턴스를 생성하는데, 원본과 1비트밖에 다르지 않다. 인스턴스를 생성하는 작업으로 BigInteger의 크기에 비례해 시간과 공간 비용이 든다. 반대로 `BitSet`은 BigInteger처럼 임의의 긴 비트 시퀀스를 나타낼 수 있는 점은 같지만 변경이 가능하다는 점이 다르다. BitSet은 상수 시간에 100만 비트 인스턴스의 비트 하나 상태를 바꿀 수 있는 메소드를 제공한다.

```
BigSet moby = ...;
moby.filp(0);
```
원하는 객체를 완성하기까지의 단계가 많고 중간 단계에서 만들어진 객체들이 모두 버려진다면 성능 문제가 불거질 수 있다. 이 문제에 대한 해결법으로는 내부에 가변 동반 클래스(companion class)를 둠으로써 다단계 연산 속도를 높여주는 것이다. 예를 들면 String의 가변 동반 클래스로 `StringBuilder`가 있다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,86 @@
+    public Task create(Task task) {
+        final Long id = generateId();
+
+        Task createdTask = Task.builder()
```
여기에도 final을 쓸 수 있겠습니다.
```suggestion
        final Task createdTask = Task.builder()
```

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,86 @@
+        return createdTask;
+    }
+
+    /**
+     * 기존의 할 일을 아이디로 찾아 수정합니다.
+     * @param id 수정할 할 일 아이디
+     * @param task 수정할 할 일 데이터
+     * @return 수정 완료된 할 일
+     */
+    public Task update(Long id, Task task) {
+        Task foundTask = findById(id)
+                .orElseThrow(() -> new IllegalArgumentException(id + "에 해당하는 할 일을 찾지 못해 업데이트를 수행할 수 없습니다"));
```
`@throws`를 JavaDoc에서 절대 빠뜨리지 않도록 주의해 주세요.

```suggestion
    /**
     * 기존의 할 일을 아이디로 찾아 수정합니다.
     * @param id 수정할 할 일 아이디
     * @param task 수정할 할 일 데이터
     * @throws IllegalArgumentException 할 일을 찾지 못한 경우
     * @return 수정 완료된 할 일
     */
    public Task update(Long id, Task task) {
        Task foundTask = findById(id)
                .orElseThrow(() -> new IllegalArgumentException(id + "에 해당하는 할 일을 찾지 못해 업데이트를 수행할 수 없습니다"));
```

johngrib reviewed (commented):
훌륭하십니다. Approve를 드리고 싶지만 추가 과제를 더 드려봤어요.

eunhasoo committed:
Docs: 예외 정보 설명 추가

eunhasoo committed:
Refactor: Task를 불변 클래스로 변경

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -69,11 +69,16 @@ public Task create(Task task) {
 
-        foundTask.updateTitle(task.getTitle());
-        return foundTask;
+        final Task updatedTask = Task.builder()
+                .id(foundTask.getId())
+                .title(task.getTitle())
+                .build();
```
Task를 불변으로 만들었지만 Task를 업데이트하는 기능은 Task 바깥에 나와 있군요. 이렇게 하면 Task를 업데이트하는 곳들에서 전부 제각각 자신만의 방식으로 Task를 생성하게 되기 때문에 시간이 흐를수록 엉망진창이 되어 갑니다.

그래서 불변 개념으로 Task를 만들게 되면 이 방법을 써 보세요.

Task 에 update 같은 이름을 갖는 메소드를 하나 만들어주고(더 나은 이름이 있다면 그걸 써도 됩니다. 예를 들어 assoc이라던가.. . with라던가... 아니면 그 외의 다양한 옵션들을 고민해 보세요), 이 메소드 안에서 실제로는 update를 하지 말고 new Task 를 생성해서 리턴하게 해보세요. 불변성을 지킨다면 setter를 만든다 해도(사실 불변 객체는 set~ 같은 이름의 메소드를 갖지 않도록 디자인하긴 하지만) 실제로는 set을 하지 않고 해당 값을 갖는 new Task를 만들어서 리턴하게 됩니다.

eunhasoo replied:
> 이렇게 하면 Task를 업데이트하는 곳들에서 전부 제각각 자신만의 방식으로 Task를 생성하게 되기 때문에 시간이 흐를수록 엉망진창이 되어 갑니다.

그렇네요. create와 로직이 거의 같게 되어버려서 뭔가 찜찜했어요 ㅋㅋㅋㅋ 객체를 새로 생성하는 update 메소드를 내부에 두면 되는 거였는데 조금 어렵게 생각했던 것 같아요. 모든 프로퍼티를 받는 것과 고민하다가 title만 받아서 처리하도록 수정해보았습니다. 실무에서 이런 식으로 불변 객체를 자주 사용하는지도 궁금합니다!

johngrib replied:
대부분의 경우 실무에서는 잘 쓰지 않을 겁니다. 보통은 그냥 상태를 바꾸죠. 하지만 불변성을 중시하는 조직에서 일하게 된다면 다들 이런 개념을 당연하다는 듯이 생각하고 말하고 코딩할 때에도 쓸 거에요. 그래서 이런 개념도 알아둬야 합니다.

eunhasoo replied:
그렇군요. 종립님 덕분에 이번에 불변 클래스에 대해서 조사하고 구현도 해보고 좋은 공부가 되었습니다ㅎㅎ 감사합니다

johngrib replied:
mutable을 기본으로 삼고 있는 Java를 사용한다 하더라도 immutable에 대해 공부해두는 건 좋은 경험이 될 수 있습니다. immutable을 기본으로 삼고 있는 다른 JVM 언어를 잠시 둘러보고 오는 것도 괜찮은 공부가 될 수 있어요. 이 글은 제가 작성한 것인데, 여유가 있을 때 한번쯤 읽어보시기를 권합니다. Clojure 언어의 Vector 구현과 Java 언어의 ArrayList 구현을 다루는 내용입니다. 

https://johngrib.github.io/wiki/clojure/study/vector/

eunhasoo committed:
Refactor: 코드 리뷰 반영

- 변경 처리를 클래스 내부에 갖도록 이동
- 입력값 검증 추가

eunhasoo committed:
Fix: 검증시 title에 this 키워드 삭제

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -18,6 +20,18 @@ public String getTitle() {
+     * @throws IllegalArgumentException title이 없을 경우
+     * @return 변경된 title을 갖는 새로운 할 일
+     */
+    public Task withNewTitle(String title) {
+        Assert.notNull(this.title, "title은 빈 값일 수 없습니다");
+        return new Task(this.id, title);
+    }
```
Assert까지 넣으시다니! 너무 잘 작업하셨습니다.

johngrib reviewed (approved):
LGTM

johngrib commented:
한 주간 노력 많으셨습니다. 다음주도 열심히 노력하실 거라 생각해요. 틈틈이 JLS 문서를 읽는 것도 잊지 않으시면 더 보람찬 한 주가 될 거라 생각해요. 다음주도 화이팅!
